deep merge property matchers in snapshot objects

swcisel opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory. -->

<!-- Please remember to update CHANGELOG.md in the root of the project if you have not done so. -->

## Summary

Correct an issue where snapshots that were tested against property matchers would not be merged with the rest of the object being tested.

## Test plan

All tests pass.

swcisel commented:
I believe this fixes the issue mentioned in #6455. This is my first PR on this project, so I'll happily take any advice or guidance from the community.

mattphillips commented on packages/jest-snapshot/src/index.js:
```diff
@@ -48,6 +48,27 @@ const cleanup = (hasteFS: HasteFS, update: SnapshotUpdateState) => {
+    Object.keys(source).forEach(key => {
+      if (isObject(source[key]) && !source[key].$$typeof) {
+        if (!(key in target))
+          Object.assign(output, {[key]: source[key]});
```
I'm not sure where `output` is defined? Should this not be:

```js
Object.assign(mergedOutput, {[key]: source[key]});
```

swcisel replied:
Thanks for catching that, it was late when I submitted this and I overlooked it.

mattphillips commented on packages/jest-snapshot/src/index.js:
```diff
@@ -48,6 +48,27 @@ const cleanup = (hasteFS: HasteFS, update: SnapshotUpdateState) => {
+        if (!(key in target))
+          Object.assign(output, {[key]: source[key]});
+        else
+          output[key] = deepMerge(target[key], source[key]);
```
Same as above

mattphillips commented on packages/jest-snapshot/src/index.js:
```diff
@@ -48,6 +48,27 @@ const cleanup = (hasteFS: HasteFS, update: SnapshotUpdateState) => {
+        else
+          output[key] = deepMerge(target[key], source[key]);
+      } else {
+        Object.assign(output, {[key]: source[key]});
```
Same as above

SimenB commented on CHANGELOG.md:
```diff
@@ -14,6 +14,7 @@
 - `[expect]` `toEqual` no longer tries to compare non-enumerable symbolic properties, to be consistent with non-symbolic properties. ([#6398](https://github.com/facebook/jest/pull/6398))
 - `[jest-util]` `console.timeEnd` now properly log elapsed time in milliseconds. ([#6456](https://github.com/facebook/jest/pull/6456))
 - `[jest-mock]` Fix `MockNativeMethods` access in react-native `jest.mock()` ([#6505](https://github.com/facebook/jest/pull/6505))
+- `[jest-snapshot]` Correctly merge property matchers with the rest of the snapshot in `toMatchSnapshot`. ([#6455](https://github.com/facebook/jest/issues/6455))
```
this should refer to this PR, not the issue

SimenB commented on packages/jest-snapshot/src/index.js:
```diff
@@ -48,6 +48,27 @@ const cleanup = (hasteFS: HasteFS, update: SnapshotUpdateState) => {
+  return (item && typeof item === 'object' && !Array.isArray(item));
+}
+
+function deepMerge(target, source) {
```
I'd very much prefer if we could find some deep merge on npm. If that's not possible, this should be moved to a util file and receive quite extensive unit tests 🙂

swcisel replied:
That's a very good point. It doesn't make a lot of sense to clutter this file up with a general-purpose merge function. I'll take a look at this tonight.

SimenB commented:
>@SimenB @thymikee @rickhanlonii do we already have a deep merge function defined anywhere else in the codebase that can be reused?

Nope, we only have shallow merge (and that uses `Object.assign`)

swcisel commented:
I've moved the deepMerge function to jest-snapshot's utils, and added a simple test. Before I add more thorough tests, would anyone mind weighing in on whether I'm headed in the right direction? I also did some digging into existing deep merge solutions, but I think keeping it local is our best bet.

SimenB commented:
I think this is a good starting point, yeah 🙂

rickhanlonii commented on packages/jest-snapshot/src/__tests__/utils.test.js:
```diff
@@ -198,3 +199,15 @@ test('serialize handles \\r\\n', () => {
+    const mergedOutput = deepMerge(target, propertyMatchers);
+
+    expect(mergedOutput).toStrictEqual({data: {bar: 'bar', foo: matcher}});
+    expect(target).toStrictEqual({data: {bar: 'bar', foo: 'foo'}});
```
@swcisel I updated your tests here, does this make sense?

swcisel replied:
Looks good to me. thanks!

rickhanlonii commented on packages/jest-snapshot/src/index.js:
```diff
@@ -53,6 +53,12 @@ const toMatchSnapshot = function(
+    throw new Error(
+      'Property matchers must be an object.\n\nTo provide a snapshot test name without property matchers, use: toMatchSnapshot("name")',
+    );
+  }
```
Added the checks here to close https://github.com/facebook/jest/issues/6581

rickhanlonii commented:
@SimenB what do you think now?
